pkg/managed/serviceaccount_test.go

```diff
@@ -0,0 +1,237 @@
+	err := CreateServiceAccounts("testBinding", imagePullSecrets, managedClusters, managedConnections)
+	if err != nil {
+		t.Fatal(fmt.Sprintf("can't create service account: %v", err))
+	}
+
+	// Validate the service accounts created for the "local" cluster
+	managedClusterConnection := managedConnections["local"]
+	serviceAccounts, err := managedClusterConnection.ServiceAccountLister.List(k8slabels.Everything())
+	if err != nil {
+		t.Fatal(fmt.Sprintf("can't list service account for cluster local: %v", err))
+	}
+	assert.Equal(t, 2, len(serviceAccounts), "2 service accounts were expected for cluster local")
+	for _, sa := range serviceAccounts {
+		assertCreateAppServiceAccounts(t, sa, "local")
+	}
+
+	// Validate the service accounts created for the "cluster-1" cluster
+	managedClusterConnection = managedConnections["cluster-1"]
+	serviceAccounts, err = managedClusterConnection.ServiceAccountLister.List(k8slabels.Everything())
+	if err != nil {
+		t.Fatal(fmt.Sprintf("can't list service account for cluster cluster-1: %v", err))
+	}
+	assert.Equal(t, 1, len(serviceAccounts), "1 service account was expected for cluster cluster-1")
+	for _, sa := range serviceAccounts {
+		assertCreateAppServiceAccounts(t, sa, "cluster-1")
+	}
+
+	// Update the imagePullSecret for service account we created in the "cluster-1" cluster.
+	serviceAccounts[0].ImagePullSecrets[0].Name = ""
+	_, err = managedClusterConnection.KubeClient.CoreV1().ServiceAccounts("ns-test3").Update(context.TODO(), serviceAccounts[0], metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatal(fmt.Sprintf("can't update service account for cluster cluster-1: %v", err))
+	}
+
+	// Call CreateServiceAccounts again, the update code will be executed to make the service account right.
+	err = CreateServiceAccounts("testBinding", imagePullSecrets, managedClusters, managedConnections)
+	if err != nil {
+		t.Fatal(fmt.Sprintf("can't create service account: %v", err))
+	}
+
+	// Validate the service account was updated for the "cluster-1" cluster
+	serviceAccounts, err = managedClusterConnection.ServiceAccountLister.List(k8slabels.Everything())
+	if err != nil {
+		t.Fatal(fmt.Sprintf("can't list service account for cluster cluster-1: %v", err))
+	}
+	assert.Equal(t, 1, len(serviceAccounts), "1 service account was expected for cluster cluster-1")
+	for _, sa := range serviceAccounts {
+		assertCreateAppServiceAccounts(t, sa, "cluster-1")
+	}
+}
+
+// TestCreateVMIServiceAccounts tests creation of service accounts when the binding name is 'system'
+// GIVEN clusters which do not have service accounts for each namespace of a VMI binding
+//  WHEN I call CreateServiceAccounts
+//  THEN there should be a service account created for each namespace of a VMI binding
+//   AND that service account should have the expected values
+func TestCreateVMIServiceAccounts(t *testing.T) {
+	assert := assert.New(t)
+
+	// Setup the needed structures to test creating service accounts for a VMI binding.
```

Comment:
Just an overall comment, and I am not saying you need to change this.  It would be better if the support functions, like createAppNamespaces, and getManagedConnections were data driven and did not have hard coded cluster names and namespace names, which makes them inter-dependent.  Also, if a map of namespace->cluster name was declared here and passed in then it would be cleared what was being validated.